Added classes, interfaces, toString(), hashCode(), equals().

NemoLABA opened the pull request:
Added classes, toString(), hashCode(), equals(). Added an enum for ExtendedWarranty along with it's own interface. Also added  interfaces for the vehicles and sales packages called IVehicleMove and IWarranty.

NemoLABA committed:
Testing commits, added package names and a Person class for better clarity.

NemoLABA committed:
Testing commits, added package names and a Person class for better clarity.

NemoLABA committed:
Testing commits, added package names and a Person class for better clarity.

NemoLABA committed:
Testing commits, added package names and a Person class for better clarity.

NemoLABA committed:
Testing commits, added package names and a Person class for better clarity.

NemoLABA committed:
Testing commits, added package names and a Person class for better clarity.

ankochergin commented on .idea/uiDesigner.xml:
add .idea folder to the ignore

NemoLABA replied:
Added .idea to ignore

ankochergin commented on src/payment/CreditCard.java:
```diff
@@ -0,0 +1,57 @@
+package payment;
+
+public class CreditCard {
+    private String cardNumber;
```
```suggestion
public class CreditCard {

    private String cardNumber;
```
please add an empty line between the class declaration and first method/variable

NemoLABA replied:
Corrected formatting

ankochergin commented on src/payment/CreditCard.java:
```diff
@@ -0,0 +1,57 @@
+    private String cardNumber;
+    private String cardHolderName;
+    private String expirationDate;
+    private String CVV;
```
only final variables can be all-uppercase, rename it to cvv or make it final (and remove setters). I think card number, cvv can be final and generated randomly on creation

NemoLABA replied:
made final and save with import java.security.SecureRandom.

ankochergin commented on src/people/Employee.java:
```diff
@@ -0,0 +1,53 @@
+    private int employeeId;
+    private String department;
+
+    // Constructor with all fields including those from the base class
```
remove comments

NemoLABA replied:
removed comments

ankochergin commented on src/people/Employee.java:
```diff
@@ -0,0 +1,53 @@
+package people;
+import java.util.Objects;
+
+public class Employee extends Person {
+    private int employeeId;
```
```suggestion
public class Employee extends Person {

    private int employeeId;
```

NemoLABA replied:
Fixed formatting.

ankochergin commented on src/people/Employee.java:
```diff
@@ -0,0 +1,53 @@
+        this.department = department;
+    }
+
+    // Optional: Override the toString method if you want to include Employee-specific details
+    @Override
+    public String toString() {
+        return super.toString() + "\nEmployee ID: " + this.employeeId + "\nDepartment: " + this.department;
+    }
+
+    // You may also override hashCode and equals if you want to include Employee-specific fields in the comparison
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), employeeId, department);
+    }
+
```
remove comments

NemoLABA replied:
Removed comments

ankochergin commented on src/people/Person.java:
```diff
@@ -0,0 +1,61 @@
+        return this.city;
+    }
+
+    public void setCity(String city) {
+        this.city = city;
+    }
+
+
+
+    @Override
+    public String toString() {
+        return this.name + "\n  " + this.address + " " + this.city;
+    }
```
```suggestion
    public void setCity(String city) {
        this.city = city;
    }

    @Override
    public String toString() {
        return this.name + "\n  " + this.address + " " + this.city;
    }
```

NemoLABA replied:
Fixed formatting.

ankochergin commented on src/vehicle/Car.java:
```diff
@@ -0,0 +1,65 @@
+    private String make;
+    private String model;
+    private int year;
+    private String VIN;    // Vehicle Identification Number
```
same as for CVV in CreditCard class

NemoLABA replied:
made final, randomly generated, and secure with import StringBuilder.

ankochergin replied:
I think VIN is still not final, I don't see any random here, could you please double-check?

ankochergin commented on src/vehicle/Gas.java:
```diff
@@ -0,0 +1,74 @@
+package vehicle;
+
+public final class Gas extends Car implements IVehicleMove {
+    private static final String VEHICLE_TYPE = "Gas"; // Final variable
+    private static int totalGasVehicles; // Static variable
+
+    private final String engineType;     // Final variable for demonstration
+    private String engine;               // Ex: 5.4L
+    private String drive;                // Ex: FWD, RWD, AWD
```
```suggestion
public final class Gas extends Car implements IVehicleMove {

    private static final String VEHICLE_TYPE = "Gas"; // Final variable
    private static int totalGasVehicles; // Static variable
    private final String engineType;     // Final variable for demonstration
    private String engine;               // Ex: 5.4L
    private String drive;                // Ex: FWD, RWD, AWD
```

NemoLABA replied:
Fixed formatting.

ankochergin commented on src/cardealership/CarDealership.java:
```diff
@@ -0,0 +1,4 @@
+package cardealership;
+
+public class CarDealership {
+}
```
store here lists of all your vehicles, employees, clients, add some addEmployee, addCustomer, addVehicleToStock(vehicle), sellVehicle(employee, customer, vehicle) etc

NemoLABA replied:
Created all the requested logic in CarDealership.

ankochergin reviewed (commented):
Add any class to store sales, it's should be the main feature in your app

NemoLABA committed:
Delete .idea directory

NemoLABA committed:
Updated CarDealership with new logic to AddVehicle, sellVehicle, addEmployee, removeEmployee. Added formatting fixes along with other changes.

NemoLABA committed:
Updated CarDealership with new logic to AddVehicle, sellVehicle, addEmployee, removeEmployee. Added formatting fixes along with other changes.

NemoLABA committed:
Testing commits, added package names and a Person class for better clarity.

ankochergin commented on src/cardealership/CarDealership.java:
```diff
@@ -0,0 +1,103 @@
+        customers.add(customer);
+    }
+
+    public void printSaleStatus(boolean saleStatus, CarSalesman salesman, Customer customer) {
+        if (saleStatus) {
+            System.out.println("Congratulations, the vehicle is yours!");
+            System.out.println("Salesman's lifetime sales: " + salesman.getLifetimeSales());
+            System.out.println("Customer's vehicle's purchased: " + customer.getVehiclesPurchased());
+        } else {
+            System.out.println("Vehicle not in stock, sale not completed.");
+        }
+    }
```
I think this method can be private and you can call it not from your main, but from your sellVehicle method. So, if sale was completed it will print success info, otherwise unsuccessful method

NemoLABA replied:
modified SellVehicle to include logger info for success, if fails then it calls printSaleStatus with logger level error.

ankochergin commented on src/cardealership/CarDealership.java:
```diff
@@ -0,0 +1,103 @@
+
+public class CarDealership {
+    private static final Logger logger = LogManager.getLogger(CarDealership.class);
+    private List<Gas> gasCars = new ArrayList<>();
+    private List<Electric> electricCars = new ArrayList<>();
+    private List<Motorcycle> motorcycles = new ArrayList<>();
+    private List<Car> allVehicles = new ArrayList<>();
```
you can store only 1 list allVehicles and to getElectricCars/getMotorcycles/getGasCars you can go through the list of all cars and check it's VEHICLE_TYPE

ankochergin replied:
it'll help us to avoid data duplication

NemoLABA replied:
made all vehicles into 1 list allVehicles, created a mthod to getGasCars, getElectricCars, getMotorcycleCars.

ankochergin commented on src/Main.java:
```diff
@@ -1,5 +1,50 @@
 public class Main {
-    public static void main(String[] args) {
-        System.out.println("Hello world!");
+    private static final Logger logger = LogManager.getLogger(Main.class);
```
logger is final, so please rename it to LOGGER in all your classes to follow naming conventions

ankochergin replied:
also please change all of your System.out.println to LOGGER and add some more logging with different levels

NemoLABA replied:
since this is meant for the next PR task  i'm gonna fix these changes and merge this PR and post the new one with this change included.

ankochergin commented on src/people/CarSalesman.java:
```diff
@@ -0,0 +1,63 @@
+
+import java.util.Objects;
+
+public final class CarSalesman extends Employee {
+    private static final Logger logger = LogManager.getLogger(CarSalesman.class);
+    private double lifetimeSales;
+    private double customerSatisfactionRating;
```
```suggestion
public final class CarSalesman extends Employee {

    private static final Logger logger = LogManager.getLogger(CarSalesman.class);
    
    private double lifetimeSales;
    private double customerSatisfactionRating;
```

ankochergin replied:
please add the same empty lines in all classes to make it easier to read

NemoLABA replied:
Added empty lines for better formatting and readability.

ankochergin commented on src/Main.java:
```diff
@@ -1,5 +1,50 @@
-    public static void main(String[] args) {
-        System.out.println("Hello world!");
+    private static final Logger logger = LogManager.getLogger(Main.class);
+    public static void main(String[] args) throws InvalidNameException {
```
please don't throw exception from main, add try-catch in main for the method which throws an exception

NemoLABA replied:
Added the try catch to the method and not main.

ankochergin commented on lib/log4j2.xml:
```diff
@@ -0,0 +1,13 @@
+        </Console>
+    </Appenders>
+    <Loggers>
+        <Root level="info" additivity="false">
+            <AppenderRef ref="console" />
+        </Root>
+    </Loggers>
+</Configuration>
```
add logging to the file

NemoLABA replied:
Will add to new PR for exceptions and logger.

ankochergin replied:
please add it in this PR

NemoLABA committed:
Added exceptions to the method in main,fixed formatting issues, changed lists of all, gas, electric and motorcycle to just all and grabbing each vehicle from the all list through VEHICLE_TYPE.

NemoLABA committed:
Adjusted Logger to be LOGGER since it's final.

NemoLABA committed:
Added More Exceptions, Class logic and try catch in main.

Added 5 total Exceptions, more logic in classes to handle the exceptions. For main I started using try catch for the objects and object methods.

NemoLABA committed:
Replaced all System,out.println with LOGGER.info

Replaced all System.out.println with LOGGER.info, used parameterized logging for better managing of log levels.

ankochergin commented on src/exceptions/InvalidDepartmentException.java:
```diff
@@ -0,0 +1,7 @@
+package exceptions;
+
+public class InvalidDepartmentException extends Exception {
+    public InvalidDepartmentException(String message) {
+        super(message);
+    }
+}
```
```suggestion
public class InvalidDepartmentException extends Exception {

    public InvalidDepartmentException(String message) {
        super(message);
    }
    
}
```

NemoLABA replied:
Fixed formatting

ankochergin commented on src/people/Customer.java:
```diff
@@ -0,0 +1,67 @@
+    private static final Logger LOGGER = LogManager.getLogger(Customer.class);
+
+    private String phoneNumber;
+    private int vehiclesPurchased;
```
I think it's some kind of data duplication, is it possible to remove it?

NemoLABA replied:
I created a method to increase vehicles purchased and I call it in main so I will have to refactor some stuff. Will use collections and possibly a tree map

ankochergin commented on src/sales/IWarranty.java:
```diff
@@ -0,0 +1,6 @@
+package sales;
+
+public interface IWarranty {
+    void extendWarranty(WarrantyType type);
+    boolean isWarrantyValid();
+}
```
```suggestion
public interface IWarranty {

    void extendWarranty(WarrantyType type);
    boolean isWarrantyValid();
    
}
```

NemoLABA replied:
fixed formatting

ankochergin commented on src/sales/WarrantyType.java:
```diff
@@ -0,0 +1,17 @@
+package sales;
+
+public enum WarrantyType {
+    BASIC(500),
```
```suggestion
public enum WarrantyType {

    BASIC(500),
```

NemoLABA replied:
fixed formatting

ankochergin commented on src/Main.java:
```diff
@@ -1,5 +1,72 @@
+import java.util.List;
+
+
 public class Main {
+    private static final Logger LOGGER = LogManager.getLogger(Main.class);
     public static void main(String[] args) {
```
```suggestion
public class Main {

    private static final Logger LOGGER = LogManager.getLogger(Main.class);
    
    public static void main(String[] args) {
```

NemoLABA replied:
fixed formatting

NemoLABA committed:
Created lib.iml

Since I have the Log4j2.xml file in my lib I thought I might need log4j core and API in it.
